### ISSUE TITLE
rpc connection messages report connections dropped - not clear if connections came back up

### ISSUE THREAD

**Issue by rachlenko** [scylladb/scylla#973]
In the logs I see connections drop between DC's yet it is not clear if the connection came up back or it is still down

```
<30>Mar  2 05:13:31 XXXXX scylla[11576]:  [shard 0] rpc - client 10.10.122.74: client connection dropped: read: Connection timed out
```

```
<30>Mar  2 04:52:51 XXXX scylla[32115]:  [shard 17] rpc - client 10.20.82.88: client connection dropped: read: Connection timed out
<30>Mar  2 04:57:41 XXXX scylla[32115]:  [shard 17] rpc - client 10.20.83.174: client connection dropped: read: Connection timed out
<30>Mar  2 05:00:53 XXXX scylla[32115]:  [shard 16] rpc - client 10.20.82.179: client connection dropped: read: Connection timed out
<30>Mar  2 05:00:55 XXXX scylla[32115]:  [shard 17] rpc - client 10.20.82.179: client connection dropped: read: Connection timed out
<30>Mar  2 05:17:37 XXXX scylla[32115]:  [shard 13] rpc - client 10.20.82.102: client connection dropped: read: Connection timed out
<30>Mar  2 05:24:35 XXXX scylla[32115]:  [shard 17] rpc - client 10.20.82.118: client connection dropped: read: Connection timed out
```

**Comment by tzach** [comment]
@gleb-cloudius can you take a look?

**Comment by gleb-cloudius** [comment]
On Wed, Mar 02, 2016 at 02:31:18AM -0800, Tzach Livyatan wrote:
> @gleb-cloudius can you take a look?
> 
There is no one global connection status between two nodes. A lot of
connections are opened between each pair of nodes (different message
types go over different connection and all is per shard). There is no
log point on successful connection.

--
			Gleb.

**Comment by dorlaor** [comment]
On Wed, Mar 2, 2016 at 12:56 PM, Gleb Natapov <notifications@github.com>
wrote:

> On Wed, Mar 02, 2016 at 02:31:18AM -0800, Tzach Livyatan wrote:
> > @gleb-cloudius can you take a look?
> >
> There is no one global connection status between two nodes. A lot of
> connections are opened between each pair of nodes (different message
> types go over different connection and all is per shard). There is no
> log point on successful connection.
>

Trying to decypher your answer - are you saying that the above logs
do not necessarily indicate there is a problem?

Do we have another mechanism to track the rpc status?
Maybe our collectd metric? We could add a REST api for it too

**Comment by gleb-cloudius** [comment]
On Wed, Mar 02, 2016 at 03:09:29AM -0800, Dor Laor wrote:
> On Wed, Mar 2, 2016 at 12:56 PM, Gleb Natapov <notifications@github.com>
> wrote:
> 
> > On Wed, Mar 02, 2016 at 02:31:18AM -0800, Tzach Livyatan wrote:
> > > @gleb-cloudius can you take a look?
> > >
> > There is no one global connection status between two nodes. A lot of
> > connections are opened between each pair of nodes (different message
> > types go over different connection and all is per shard). There is no
> > log point on successful connection.
> >
> 
> Trying to decypher your answer - are you saying that the above logs
> do not necessarily indicate there is a problem?
> 
They indicate that there were problems during _a_ connection, but there is
no _the_ connection between two nodes. We have many independent tcp streams between
each pair of nodes. Log indicates that some (or all) had problems and
were closed. Next communication attempt will try to open another
connection. Since we do not log when connection succeeds there is no way to
tell when connectivity is restored short of not seeing errors in the log
any more.


> Do we have another mechanism to track the rpc status?
RPC does not have such status. It is a thin layer on top of tcp.

> Maybe our collectd metric? We could add a REST api for it too
> 
> 
It has some JMX counters, but I do not see anything helpful that can
indicate that connectivity between two nodes is restored better then
simple ping. The counters seams to be not per dst (better ask Amnon).

--
			Gleb.

**Comment by slivne** [comment]
Are this logged because we tried to send something over the connection and
then the connection was dropped ?

Or is this logged because keep alive was not received ?

We need some more info on when this is logged - and if this can happen in
multiple cases for multiple reasons then I think that we need to add this
to the log (if it doesn't exist)

If this is logged and a message with data was dropped - do we retry to send
the message over a new connection or simply drop the message (I know this
is a messaging_service question - but I guess you know the answer)?

**Comment by avikivity** [comment]
The errors happen when a TCP SYN was sent, but not ACKed.  Could it be that the node is down (powered off) when the connection was initiated?

**Comment by slivne** [comment]
No the servers are all up - there are parallel connections

**Comment by gleb-cloudius** [comment]
On Mon, Mar 07, 2016 at 05:52:51AM -0800, slivne wrote:
> Assigned #973 to @gleb-cloudius.
> 
As I explained there is no such things as "the connection". This cannot be fixed.

--
			Gleb.

**Comment by tgrabiec** [comment]
@gleb-cloudius There is no single connection, but there are (per-shard) connections. Could we track/log connection state per-shard?

**Comment by gleb-cloudius** [comment]
On Mon, Mar 07, 2016 at 07:19:00AM -0800, Tomasz Grabiec wrote:
> @gleb-cloudius There is no single connection, but there are (per-shard) connections. Could we track/log connection state per-shard?
> 
There is more then one connection per shard.

--
			Gleb.

**Comment by tgrabiec** [comment]
2016-03-07 16:39 GMT+01:00 Gleb Natapov <notifications@github.com>:

> On Mon, Mar 07, 2016 at 07:19:00AM -0800, Tomasz Grabiec wrote:
> > @gleb-cloudius There is no single connection, but there are (per-shard)
> connections. Could we track/log connection state per-shard?
> >
> There is more then one connection per shard.
>

Right. Each of them can be uniquely identified, could we track state of
each such connection?

**Comment by gleb-cloudius** [comment]
On Mon, Mar 07, 2016 at 08:32:20AM -0800, Tomasz Grabiec wrote:
> 2016-03-07 16:39 GMT+01:00 Gleb Natapov <notifications@github.com>:
> 
> > On Mon, Mar 07, 2016 at 07:19:00AM -0800, Tomasz Grabiec wrote:
> > > @gleb-cloudius There is no single connection, but there are (per-shard)
> > connections. Could we track/log connection state per-shard?
> > >
> > There is more then one connection per shard.
> >
> 
> Right. Each of them can be uniquely identified, could we track state of
> each such connection?
> 
This is the same as asking can you determine tcp connectivity status to
host A in tcp stack. What does this mean? RPC has no idea how much
connections upper layer wants to open to a host A, so such question at this
level does not make sense. Messaging server expects to have two
connections per shard per active destination. But it rightfully does not
care about "state" whatever that means. If connection does not exists
(which means it was never attempted) it will open one and send data. We
can add possibility to query if connection is in a connected state or
still connection and query it in rest api somehow from messaging
service, so the "state" of connection will be "does not exist" (means
nothing about connectivity), "connected" and "connecting". Of course
more or less same information is available from netstat sans per shard,
but shard is not important for connectivity issues.

--
			Gleb.

**Comment by slivne** [comment]
adding support for setting keep alive for connections solved the issue

**Comment by tzach** [comment]
@gleb-cloudius @slivne which commit fix this issue?

**Comment by tzach** [comment]
fixed by e228ef1bd97c470e9aaa4063775c7f66d630b473
